src/profiler.ts

```diff
@@ -13,57 +13,203 @@
-// TODO: add stop() method to stop profiling.
+// Converts a profile to a compressed, base64 encoded string.
+function profileBytes(p: perftools.profiles.IProfile): Promise<string> {
+  let time = new Date((p.timeNanos as number) / 1000000);
+  let pwriter = perftools.profiles.Profile.encode(p);
+  let buffer = new Buffer(pwriter.finish());
+  return new Promise<string>((resolve, reject) => {
+    zlib.gzip(buffer, (err: Error, gzBuf: Buffer) => {
+      err ? reject(err) : resolve(gzBuf.toString('base64'));
+    });
+  });
+}
+
+// Service object for communicating with Stackdriver Profiler API.
+class ProfilerServiceObject extends common.ServiceObject {
+  constructor(options: AuthenticationConfig) {
+    options = common.util.normalizeArguments(null, options);
+    const config = {
+      baseUrl: API,
+      scopes: ['https://www.googleapis.com/auth/monitoring.write'],
+      packageJson: pjson,
+    };
+    super({parent: new common.Service(config, options), baseUrl: '/'});
+  }
+}
+
+// Polls Stackdriver Profiler server for instructions on behalf of a task and
+// collects and uploads profiles as requested.
 export class Profiler {
-  config: ProfilerConfig;
-  logger: Logger;
-  service: Service;
+  private config: ProfilerConfig;
+  private logger: Logger;
+  private serviceObject: ServiceObject;
+  private profileTypes: string[];
+  private continueProfiling: boolean;
+
+  // Public for testing.
+  timeProfiler: TimeProfiler|undefined;
 
   constructor(config: ProfilerConfig) {
     this.config = config;
-
-    this.service = new common.Service(
-        {
-          baseUrl: 'https://cloudprofiler.googleapis.com/v2',
-          scopes: ['https://www.googleapis.com/auth/monitoring.write'],
-        },
-        this.config);
-
+    this.serviceObject = new ProfilerServiceObject(config);
     this.logger = new common.logger({
       level: common.logger.LEVELS[config.logLevel as number],
       tag: pjson.name
     });
+
+    // TODO: enable heap profiling once heap-profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push('WALL');
+      this.timeProfiler = new TimeProfiler(1000);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  // Starts and endless loop to poll Stackdriver Profiler server for
+  // instructions, and collects and uploads profiles as requested.
+  // If there is a problem when collecting a profile or uploading a profile to
+  // Stackdriver Profiler, this problem will be logged at the debug level.
+  // If there is a problem polling Stackdriver Profiler for instructions
+  // on the type of profile created, an error will be thrown.
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    while (true) {
```

Comment:
This would keep a program alive forever.

Reply:
Yeah -- I wasn't certain as to a way around this. I know we need to ask the API what profile we should collect, then collect and upload the profile.

I'd considered modifying it so it would be like 
 ` ` ` 
async start(): Promise<void> {
   this.continueProfiling = true;
   while(this.continueProfiling) {
      // stuff
  }
 // more stuff
}

stop() {
  this.continueProfiling = false;
}
 ` ` ` 

But that would really only help when a user only had a few copies of the same projectId/service running, since most time is spent in createProfile() waiting for the API to indicate one should profile. And since the API only profiles once a minute per projectId/service/version combination, this wouldn't be very helpful when there are lots of jobs with the same projectId/service/version.